start work on example script to add all the needed metadata (#80)
metno/data-management-handbook: source/add_metadata_python_script_example.adoc

[[example-script-add-metadata]]
===== Example script to add metadata to a netCDF file

> The below is an example python script to add to a netCDF the metadata needed for it to be parsed by https://github.com/metno/py-mmd-tools/blob/master/py_mmd_tools/nc_to_mmd.py:

Reviewer: ```suggestion
The below is an example python script to add metadata to a netCDF-CF file. Note that we have only included the required attributes, and that the data producer should add more. This is needed for the netCDF file to be parsed into an MMD xml file by the script `scripts/nc2mmd.py` in https://github.com/metno/py-mmd-tools[py-mmd-tools]:
```

[source, python]
----
#!/usr/bin/env python

> 

Reviewer: ```suggestion
# filename.py

```

from datetime import datetime, timedelta, timezone
import sys
import netCDF4 as nc
from uuid import uuid4

with nc.Dataset(sys.argv[1], "a") as f:
    f.id = str(uuid4())

> 

Reviewer: f.history += 'isotime, Dummy User, filename.py\n' # history should already exist in the netcdf file

    f.keywords_vocabulary += 'GCMDLOC:GCMD Locations:https://gcmd.earthdata.nasa.gov/kms/concepts/concept_scheme/locations, '
    f.keywords_vocabulary += 'GEMET:INSPIRE themes:http://inspire.ec.europa.eu/theme, '
    f.keywords_vocabulary += 'NORTHEMES:GeoNorge Themes:https://register.geonorge.no/subregister/metadata-kodelister/kartverket/nasjonal-temainndeling'
    f.keywords_resource = 'GCMD:https://gcmdservices.gsfc.nasa.gov/static/kms/, GEMET:http://inspire.ec.europa.eu/theme, '
    f.keywords_resource += 'NORTHEMES:https://register.geonorge.no/subregister/metadata-kodelister/kartverket/nasjonal-temainndeling'

    f.institution = 'Norwegian Meteorological Institute'
    f.institution_short_name = 'METNO'
    f.project = 'Govermental core service'
    f.project_short_name = 'Govermental core service'

    # MMD mandatory not in ACDD
    f.metadata_status = 'Active'
    f.dataset_production_status = 'Complete'
    f.iso_topic_category = 'climatologyMeteorologyAtmosphere,environment'

    f.access_constraint = 'Open'
    f.spatial_representation = 'grid'

    f.naming_authority = 'no.met'

>     # only 1 timestep in this example

Reviewer: ```suggestion
    # only 1 timestep (e.g., one measurement) in this example
```

>     coverstart = datetime(1900, 1, 1, 0, 0)  + timedelta(44012)

Reviewer: ```suggestion
    coverstart = datetime(1900, 1, 1, 0, 0)  + timedelta(44012)  # time of first measurement
```

>     coverend = coverstart

Reviewer: ```suggestion
    coverend = coverstart  # time of last measurement
```

    f.time_coverage_start = '{:%Y-%m-%dT%H:%M:%S.%f}Z'.format(coverstart)
    f.time_coverage_end = '{:%Y-%m-%dT%H:%M:%S.%f}Z'.format(coverend)

>     f.ancillary_timeliness = '[???]'

Reviewer: ```suggestion
```

    f.date_created = '{:%Y-%m-%dT%H:%M:%S}Z'.format(createddate)

    f.date_metadata_modified = '{:%Y-%m-%dT%H:%M:%S}Z'.format(datetime.now(tz=timezone.utc))
    f.date_metadata_modified_type = 'Created'

>     f.Conventions = 'CF-1.6 for coordinates,ACDD-1.3'

Reviewer: ```suggestion
    f.Conventions = 'CF-1.6, ACDD-1.3'
```

[[example-script-add-metadata]]
===== Example script to add metadata to a netCDF file

The below is an example python script to add metadata to a netCDF-CF file. Note that we have only included the required attributes, and that the data producer should add more. This is needed to support (meta)data interoperability and can be validated by the script `scripts/nc2mmd.py` in https://github.com/metno/py-mmd-tools[py-mmd-tools], which also can parse the netCDF-CF file into an MMD xml file:
[source, python]
----
#!/usr/bin/env python

from datetime import datetime, timedelta, timezone
import sys
import netCDF4 as nc
from uuid import uuid4
import os

with nc.Dataset(sys.argv[1], "a") as f:
    f.id = str(uuid4())

    isotime_now = '{:%Y-%m-%dT%H:%M:%S}Z'.format(datetime.now(tz=timezone.utc))
    isotime_now_string = str(isotime_now)
    script_name = os.path.basename(__file__)
    
    if 'history' not in f.ncattrs():
        """ History should already exist in the netcdf file but if
        it does not, we create it here.
        """
        f.history = ""

>         print("warning: no history field in the netCDF metadata")

Reviewer: strictly speaking, I would use the standard warning package instead..

    
    f.history += isotime_now_string + ", Dummy User, " + script_name + "\n" 

>     f.collection = '[e.g. METNCS]'

Reviewer: This is better to leave out. It is not in the attributes list either, is it?
```suggestion
```

    f.title = 'lorem ipsum'

    f.summary = 'lorem ipsum '
    f.summary += 'lorem ipsum '
    f.summary += 'lorem ipsum'

    f.license = 'https://spdx.org/licenses/CC-BY-4.0(CC-BY-4.0)'

    # example keywords given here

>     f.keywords = 'GCMDSK:Earth Science > Atmosphere > Atmospheric radiation, '

Reviewer: Maybe we should describe the example dataset a bit before the script or in a comment here so people don't just copy it and use it for

Reviewer: yes - we need to say which variables that are supposed to be in the example dataset, and connect the keywords to them. This should include explicit comments in the code.

Reviewer: This is now wind speed and direction, as mentioned above. Can you update the keywords?

>     f.keywords += 'GCMDLOC:Geographic Region > Northern Hemisphere, '

Reviewer: Why did you add this? I am curious, since it will be useful to know where the confusion is coming from. In principle, the keywords should be descriptive of the data inside the dataset.
```suggestion
```

Author: is this not consistent with the geographic span of the data?  
since we have 
```
    f.geospatial_lat_min = 30.05
    f.geospatial_lat_max = 81.95
    f.geospatial_lon_min = -29.95
    f.geospatial_lon_max = 89.95
```
but maybe I misunderstand the field

Reviewer: I see. I don't think the geographic location is needed here, although it is not necessarily a problem. The important ones are GCMDSK, GEMET and NORTHEMES. I'll commit the change suggestion above.

>     f.keywords += (
>                     'GCMDPROV: Government Agencies-non-US > Norway > '
>                     'NO/MET > Norwegian Meteorological Institute, '
>                 )

Reviewer: same for this
```suggestion
```

    f.keywords += (
                    'GEMET:Meteorological geographical features, '
                    'GEMET:Atmospheric conditions, '
                )
    f.keywords += (
                    'GEMET:Oceanographic geographical features, '
                    'NORTHEMES:Weather and climate'
                )
    f.keywords_vocabulary = (
                    'GCMDSK:GCMD Science Keywords:https://gcmd.'
                    'earthdata.nasa.gov/kms/concepts/concept_scheme/'
                    'sciencekeywords, '
                )

>     f.keywords_vocabulary += (
>                     'GCMDPROV:GCMD Providers:https://gcmd.earthdata.'
>                     'nasa.gov/kms/concepts/concept_scheme/providers, '
>                 )
>     f.keywords_vocabulary += (
>                     'GCMDLOC:GCMD Locations:https://gcmd.earthdata.'
>                     'nasa.gov/kms/concepts/concept_scheme/locations, '
>                 )

Reviewer: ```suggestion
```

    f.keywords_vocabulary += (
                    'GEMET:INSPIRE themes:http://inspire.ec.europa.eu'
                    '/theme, '
                )
    f.keywords_vocabulary += (
                    'NORTHEMES:GeoNorge Themes:https://register.'
                    'geonorge.no/subregister/metadata-kodelister/'
                    'kartverket/nasjonal-temainndeling'
                )

>     f.keywords_resource = (
>                     'GCMD:https://gcmdservices.gsfc.nasa.gov/static/'
>                     'kms/, GEMET:http://inspire.ec.europa.eu/theme, '
>                 )
>     f.keywords_resource += (
>                     'NORTHEMES:https://register.geonorge.no/'
>                     'subregister/metadata-kodelister/kartverket/'
>                     'nasjonal-temainndeling'

Reviewer: `keywords_resource` is not in the attributes list either..
```suggestion
```

    f.naming_authority = 'no.met'

    # only 1 timestep (e.g., one measurement) in this example
    # time of first measurement:

>     coverstart = datetime(1900, 1, 1, 0, 0)  + timedelta(44012)

Reviewer: we should set the timezone here as well, since the "Z" below refers to utc... And maybe print as isoformat to avoid hardcoding. What if the time that a user has is not really in utc? Then an ignorant user will introduce an error.

    coverend = coverstart  # time of last measurement

>     f.time_coverage_start = '{:%Y-%m-%dT%H:%M:%S.%f}Z'.format(coverstart)

Reviewer: use another function to print the time variable directly in isoformat

    f.time_coverage_end = '{:%Y-%m-%dT%H:%M:%S.%f}Z'.format(coverend)

>     if 'created_date' in f.ncattrs() and 'created_hour' in f.ncattrs():

Reviewer: `created_date` and `created_hour` are not standard attributes. You should rather assume that you have a time variable, and use the first and the last time in that. This assumption should also be pointed out.

                    'GEMET:Atmospheric conditions, '
                )
    f.keywords += (
                    'GEMET:Oceanographic geographical features, '
                    'NORTHEMES:Weather and climate'
                )
    f.keywords_vocabulary = (
                    'GCMDSK:GCMD Science Keywords:https://gcmd.'
                    'earthdata.nasa.gov/kms/concepts/concept_scheme/'
                    'sciencekeywords, '
                )
    f.keywords_vocabulary += (
                    'GEMET:INSPIRE themes:http://inspire.ec.europa.eu'
                    '/theme, '
                )
    f.keywords_vocabulary += (
                    'NORTHEMES:GeoNorge Themes:https://register.'
                    'geonorge.no/subregister/metadata-kodelister/'
                    'kartverket/nasjonal-temainndeling'
                )

    f.naming_authority = 'no.met'

    # extracting f.time_coverage_start and f.time_coverage_end from the time variable
    coverstart = f.variables['time'][0] # time of first measurement
    coverend = f.variables['time'][-1]  # time of last measurement

    # we assume here that the time variable contains timesteps in UTC

>     try:

Reviewer: Simplify code:
```suggestion
```

Author: am not sure I understand here, you want to eliminate the check?

Reviewer: yes, we should assume it's working since we just need to show the principles

>         f.time_coverage_start = datetime.fromtimestamp(coverstart).replace(tzinfo=timezone.utc).isoformat()

Reviewer: ```suggestion
    f.time_coverage_start = datetime.fromtimestamp(coverstart).replace(tzinfo=timezone.utc).isoformat()
```

>     except:
>         warnings.warn("Warning: conversion of first timestep to datetime in ISO format failed, setting time_coverage_start as ")

Reviewer: ```suggestion
```

>     try:

Reviewer: ```suggestion
```

       f.time_coverage_end = datetime.fromtimestamp(coverend).replace(tzinfo=timezone.utc).isoformat()

>     except:
>         warnings.warn("Conversion of last timestep to datetime in ISO format failed, time_coverage_end will not be set")

Reviewer: ```suggestion
```

        """
        if 'created_date' in f.ncattrs():
            if 'created_hour' in f.ncattrs():
                try :
                    f.date_created = datetime.strptime(f.created_date+'T'+f.created_hour,'%Y%m%dT%H%M%S.%f').replace(tzinfo=timezone.utc).isoformat()
                except:
                    warnings.warn("Warning: created_date or created_hour are not in the expected format, date_created will use the current time in UTC")
                    f.date_created = isotime_now_utc
            else:
                try:
                    f.date_created = datetime.strptime(f.created_date,'%Y%m%d').replace(tzinfo=datetime.timezone.utc).isoformat()
                    warnings.warn("Warning: using current created_date and midnight time in UTC to fill the date_created field")
                except:
                    warnings.warn("Warning: created_date is not in the expected format, date_created will use the current time in UTC")
                    f.date_created = isotime_now_utc
        else:
            warnings.warn("Warning: filling the date_created field with the current time in UTC")

>             f.date_created = isotime_now_utc

Reviewer: should be simplified since it is just an example

Author: as in just setting the created date as today if not 'date_created' in f.ncattrs()?

or keeping `if 'created_date' in f.ncattrs():` and :  
 - assuming that if it's there then it's what's needed already (no check of 'created_hour') 
 - setting the created date as today if not there
 
 or something else?

Reviewer: My understanding is that you assume here that the file already exists. But does that mean that it has been created by CDO? If that's the case, it should be described before the script example. Also, if CDO adds the created_date by default then you don't need to do anything. If CDO doesn't create the attribute, then we need to add it in the example.

Another use case is if you create the nc-file from scratch. Then this needs to be explained. But I guess the CDO use case is relevant for many people at MET, so we can use that - but it needs to be clear by the introduction text before the script. The script code should be as simple as possible in order to make it easily readable, so avoid `try-except`s, `if`s, and warnings.

Author: I wasn't assuming anything specific about the creation of the file,   
created_date is not part of any standard `cdo` output afaik  
the code was inspired by created_date and created_hour as they are in the emep model run files   
I have simplified the code now

>     f.Conventions = 'CF-1.6, ACDD-1.3'

Reviewer: Check what's the latest versions...

    f.geospatial_lat_min = 30.05
    f.geospatial_lat_max = 81.95
    f.geospatial_lon_min = -29.95
    f.geospatial_lon_max = 89.95

>     f.iso_topic_category = 'climatologyMeteorologyAtmosphere,environment'

Reviewer: See https://apps.usgs.gov/thesaurus/thesaurus-full.php?thcode=15 - I don't think "environment" is correct here. At least, climatologyMeteorologyAtmosphere is good enough.
```suggestion
    f.iso_topic_category = 'climatologyMeteorologyAtmosphere'
```
